File: mainPage/informative/Training/exercises/solutions/maths3/src/MathsService.cpp
// Copyright information and license terms for this software can be
// found in the file LICENSE that is included with the distribution

#include <pv/rpcServer.h>
#include "MathsService.h"

using namespace epics::pvData;

namespace epics
{

namespace mathsService
{

// returns this service's result structure type definition.
StructureConstPtr makeResponseStructure()
{
#if 0
    FieldCreatePtr factory = getFieldCreate();

    FieldConstPtrArray fields;
    StringArray names;

    names.push_back("sum");
    fields.push_back(factory->createScalar(pvDouble));

    return factory->createStructure(names, fields);
#else
    return getFieldCreate()->
       createFieldBuilder()->add("sum", pvDouble)->createStructure();
#endif
}


// Definition of the Maths RPC service.

epics::pvData::PVStructurePtr MathsService::request(
    epics::pvData::PVStructurePtr const & pvArgument
    ) throw (pvAccess::RPCRequestException)
{   
    // Extract the two arguments.
    // Report an error by throwing a RPCRequestException
    epics::pvData::PVDoublePtr aField = pvArgument->getSubField<PVDouble>("a");
    if (!aField)
    {
        throw pvAccess::RPCRequestException(Status::STATUSTYPE_ERROR,
            "PVDouble field with name 'a' expected.");
    }

    epics::pvData::PVDoublePtr bField = pvArgument->getSubField<PVDouble>("b");
    if (!bField)
    {
        throw pvAccess::RPCRequestException(Status::STATUSTYPE_ERROR,
            "PVDouble field with name 'b' expected.");
    }


    // Create the result structure of the data interface.
    PVStructurePtr result(
        getPVDataCreate()->createPVStructure(makeResponseStructure()));

    // Fill in the value of the structure's sum field
    PVDoublePtr sumField = result->getSubField<PVDouble>("sum");
	sumField->put(aField->get() + bField->get());
    
    // return the structure.
    return result;
}

}

}

File: mainPage/informative/training/ntndarrayJava/template/ndarrayJava-template/maths/src/org/epics/trainingJava/maths/MathsService.java
/*
 * Copyright information and license terms for this software can be
 * found in the file LICENSE that is included with the distribution
 */

package org.epics.trainingJava.maths;

import org.epics.pvaccess.PVAException;
import org.epics.pvaccess.server.rpc.RPCRequestException;
import org.epics.pvaccess.server.rpc.RPCServer;
import org.epics.pvaccess.server.rpc.RPCService;
import org.epics.pvdata.factory.FieldFactory;
import org.epics.pvdata.factory.PVDataFactory;
import org.epics.pvdata.pv.Field;
import org.epics.pvdata.pv.FieldCreate;
import org.epics.pvdata.pv.PVDataCreate;
import org.epics.pvdata.pv.PVDouble;
import org.epics.pvdata.pv.PVString;
import org.epics.pvdata.pv.PVStructure;
import org.epics.pvdata.pv.ScalarType;
import org.epics.pvdata.pv.Status.StatusType;
import org.epics.pvdata.pv.Structure;

/**
 * MathsService is an example of a simple EPICS V4 Remote Procedure Call (RPC) service.
 * 
 */
public class MathsService
{
    private static final PVDataCreate pvDataCreate = 
            PVDataFactory.getPVDataCreate();
    private static final FieldCreate fieldCreate = 
            FieldFactory.getFieldCreate();

    // This service result structure type definition.

    /**
     * Implementation of RPC service.
     */
    static class MathsServiceImpl implements RPCService
    {
        public PVStructure request(PVStructure args) throws RPCRequestException
        {
            // Extract the two arguments.
            // Report an error by throwing a RPCRequestException


            // Create the result structure of the data interface.

            // Fill in the value of the structure's sum field


            // return the structure. Replace null with result
            return null;
        }
    }

    /**
     * Main is the entry point of the MathsService server side executable. 
     * @param args None
     */
    public static void main(String[] args) 
    {
        RPCServer server = new RPCServer();
        // register our service as "mathsService"
        server.registerService("mathsService", new MathsServiceImpl());
        server.printInfo();
        try {
            server.run(0);
        } catch (PVAException e) {
            System.err.println(e.getMessage());
            System.exit(1);
        }
    }
}
